File: python/src/moltchess/llm/prompt.py
from __future__ import annotations

from dataclasses import dataclass

from .types import MovePromptContext

SYSTEM_PROMPT = """You are a chess move selector for a MoltChess autonomous agent.
Each game_id is an independent chat thread. Keep separate context per game.
After the first turn, treat prior messages in the same thread as context and avoid repeating full-history analysis.
The latest FEN and legal move lists are always authoritative, even if earlier thread state appears inconsistent.
You MUST choose exactly one move from the provided legal moves only.
Reply with a single JSON object and no other text, in this shape:
{"move_san":"<SAN>","move_uci":"<UCI>"}
Use standard algebraic notation (SAN) that appears in the legal list. You may include an optional "rationale" string field; it will be ignored by the server.
If you cannot decide, pick any legal move from the list."""


@dataclass(frozen=True)
class MoveSummaryDelta:
    mode: str
    new_moves: list[str]


def _split_moves(summary: str | None) -> list[str]:
    if not summary:
        return []
    trimmed = summary.strip()
    return trimmed.split() if trimmed else []


def summarize_move_delta(previous_moves_summary: str | None, current_moves_summary: str) -> MoveSummaryDelta:
    if previous_moves_summary is None:
        return MoveSummaryDelta(mode="initial", new_moves=_split_moves(current_moves_summary))

    previous = _split_moves(previous_moves_summary)
    current = _split_moves(current_moves_summary)

    common = 0
    while common < len(previous) and common < len(current) and previous[common] == current[common]:
        common += 1

    if common != len(previous):
        return MoveSummaryDelta(mode="resync", new_moves=current)

    return MoveSummaryDelta(mode="continue", new_moves=current[common:])


def build_user_message(
    ctx: MovePromptContext,
    feedback: str | None = None,
    *,
    previous_moves_summary: str | None = None,
) -> str:
    delta = summarize_move_delta(previous_moves_summary, ctx.moves_summary)
    lines = [
        f"game_id: {ctx.game_id}",
        f"You play as: {ctx.my_color}",
    ]

    if delta.mode == "initial":
        lines.append("This is the first prompt for this game thread.")
        lines.append(f"Full move history (SAN): {ctx.moves_summary or '(start)'}")
    elif delta.mode == "resync":
        lines.append("Reset your internal game state to this authoritative snapshot.")
        lines.append(f"Full move history (SAN): {ctx.moves_summary or '(start)'}")
    elif delta.new_moves:
        lines.append("Continue the same game thread. Do not repeat prior analysis.")
        lines.append(f"New SAN moves since your last prompt: {' '.join(delta.new_moves)}")
    else:
        lines.append("No new board moves were played since your last prompt. Re-evaluate the same position.")

    lines.extend(
        [
            f"Authoritative FEN: {ctx.fen}",
            f"Legal SAN moves ({len(ctx.legal_sans)}): {', '.join(ctx.legal_sans)}",
            f"Legal UCI moves: {', '.join(ctx.legal_uci)}",
        ]
    )
    if feedback:
        lines.append(f"Previous reply was invalid: {feedback}")
        lines.append("Respond again with valid JSON and a move from the legal lists.")
    return "\n".join(lines)

File: javascript/src/llm/prompt.ts
import type { MovePromptContext } from "./types.js";

export const SYSTEM_PROMPT = `You are a chess move selector for a MoltChess autonomous agent.
Each game_id is an independent chat thread. Keep separate context per game.
After the first turn, treat prior messages in the same thread as context and avoid repeating full-history analysis.
The latest FEN and legal move lists are always authoritative, even if earlier thread state appears inconsistent.
You MUST choose exactly one move from the provided legal moves only.
Reply with a single JSON object and no other text, in this shape:
{"move_san":"<SAN>","move_uci":"<UCI>"}
Use standard algebraic notation (SAN) that appears in the legal list. You may include an optional "rationale" string field; it will be ignored by the server.
If you cannot decide, pick any legal move from the list.`;

export interface BuildUserMessageOptions {
  previousMovesSummary?: string | null;
  feedback?: string | null;
}

export interface MoveSummaryDelta {
  mode: "initial" | "continue" | "resync";
  newMoves: string[];
}

function splitMoves(summary: string | null | undefined): string[] {
  if (!summary) {
    return [];
  }
  const trimmed = summary.trim();
  return trimmed ? trimmed.split(/\s+/) : [];
}

export function summarizeMoveDelta(
  previousMovesSummary: string | null | undefined,
  currentMovesSummary: string,
): MoveSummaryDelta {
  if (previousMovesSummary === undefined || previousMovesSummary === null) {
    return { mode: "initial", newMoves: splitMoves(currentMovesSummary) };
  }

  const previous = splitMoves(previousMovesSummary);
  const current = splitMoves(currentMovesSummary);

  let common = 0;
  while (common < previous.length && common < current.length && previous[common] === current[common]) {
    common += 1;
  }

  if (common !== previous.length) {
    return { mode: "resync", newMoves: current };
  }

  return { mode: "continue", newMoves: current.slice(common) };
}

function normalizeOptions(input?: string | null | BuildUserMessageOptions): BuildUserMessageOptions {
  if (typeof input === "string" || input === null || input === undefined) {
    return { feedback: input ?? null };
  }
  return {
    previousMovesSummary: input.previousMovesSummary,
    feedback: input.feedback ?? null,
  };
}

export function buildUserMessage(
  ctx: MovePromptContext,
  input?: string | null | BuildUserMessageOptions,
): string {
  const { previousMovesSummary, feedback } = normalizeOptions(input);
  const delta = summarizeMoveDelta(previousMovesSummary, ctx.movesSummary);
  const lines = [
    `game_id: ${ctx.gameId}`,
    `You play as: ${ctx.myColor}`,
  ];

  if (delta.mode === "initial") {
    lines.push("This is the first prompt for this game thread.");
    lines.push(`Full move history (SAN): ${ctx.movesSummary || "(start)"}`);
  } else if (delta.mode === "resync") {
    lines.push("Reset your internal game state to this authoritative snapshot.");
    lines.push(`Full move history (SAN): ${ctx.movesSummary || "(start)"}`);
  } else if (delta.newMoves.length > 0) {
    lines.push("Continue the same game thread. Do not repeat prior analysis.");
    lines.push(`New SAN moves since your last prompt: ${delta.newMoves.join(" ")}`);
  } else {
    lines.push("No new board moves were played since your last prompt. Re-evaluate the same position.");
  }

  lines.push(`Authoritative FEN: ${ctx.fen}`);
  lines.push(`Legal SAN moves (${ctx.legalSans.length}): ${ctx.legalSans.join(", ")}`);
  lines.push(`Legal UCI moves: ${ctx.legalUci.join(", ")}`);

  if (feedback) {
    lines.push(`Previous reply was invalid: ${feedback}`);
    lines.push("Respond again with valid JSON and a move from the legal lists.");
  }
  return lines.join("\n");
}
